packages/core/src/com/communication.ts

```diff
@@ -1,182 +1,182 @@
+        return this.environments[envName]?.host;
+    }
+
     /**
      * Registers environments that spawned in the same execution context as the root environment.
+     * If the environment is already registered, it will not be registered again.
      */
     public registerEnv(id: string, host: Target): void {
-        if (this.DEBUG) {
-            console.debug(REGISTER_ENV(id, this.rootEnvId));
-        }
         const existingEnv = this.environments[id];
-        if (!existingEnv) {
-            this.environments[id] = { id, host } as EnvironmentRecord;
-        } else if (existingEnv.host !== host) {
-            throw new Error(DUPLICATE_REGISTER(id, 'Environment'));
+        if (existingEnv) {
+            if (existingEnv.host !== host) throw new DuplicateRegistrationError(id, 'Environment');
+
+            return;
         }
+
+        this.log?.(`Registering env ${id} at ${this.id}`);
+        this.environments[id] = { id, host };
     }
 
     /**
      * Registers local api implementation of the remote service.
      */
     public registerAPI<T extends {}>({ id }: IDTag, api: T): T {
-        if (!this.apis[id]) {
-            this.apis[id] = api;
-            this.applyApiDirectives(id, api);
-            return api;
-        } else {
-            throw new Error(DUPLICATE_REGISTER(id, 'RemoteService'));
-        }
+        if (this.apis[id]) throw new DuplicateRegistrationError(id, 'RemoteService');
+
+        this.apis[id] = api;
+        this.applyApiDirectives(id, api);
+
+        return api;
     }
 
-    public getEnvironmentContext(endPoint: ContextualEnvironment<string, EnvironmentMode, Environment[]>) {
+    public getEnvironmentContext(
+        endPoint: ContextualEnvironment<string, EnvironmentMode, Environment[]>,
+    ): string | undefined {
         return this.resolvedContexts[endPoint.env];
     }
 
-    public getEnvironmentInstanceId(envName: string, endpointType: EnvironmentMode) {
+    public getEnvironmentInstanceId(envName: string, endpointType: EnvironmentMode): string {
         return endpointType === 'single' ? envName : this.generateEnvInstanceID(envName);
     }
 
-    public getPublicPath() {
-        return this.options.publicPath;
+    /**
+     * Generate client id for newly spawned environment.
+     * todo: should be made private? or rather inlined
+     */
+    public generateEnvInstanceID(name: string): string {
+        return this.idsCounter.next(`${name}/`);
     }
 
-    public setTopology(envName: string, envUrl: string) {
-        this.topology[envName] = envUrl;
+    public getPublicPath(): string {
+        return this.options.publicPath;
     }
 
-    public subscribeToEnvironmentDispose(handler: (envId: string) => void) {
+    public subscribeToEnvironmentDispose(handler: (envId: string) => void): void {
         this.disposeListeners.add(handler);
     }
-    public unsubscribeToEnvironmentDispose(handler: (envId: string) => void) {
```

Comment:
where is this function?

Reply:
👍🏼 Yeah, got cleaned up as unused, but it was wrong on my part; I checked that it is unused both internally and externally, but externally, it should be used alongside `subscribeToEnvironmentDispose` I think